### PR TITLE
[Fabric] Fixes findNodeAtPoint when views were inverted

### LINKED ISSUE #45502: Inspector Tools issue when view are transformed in Fabric
### Description

When Views are transformed (as seen in the example with `Vertical Inversion`) and try to inspect element, wrong elements is being inspected because `findNodeAtPoint` in Fabric is not accounting for transform.

This is happening in both iOS and Android

### Steps to reproduce

1. Run iOS project with Fabric `RCT_NEW_ARCH_ENABLED=1 pod install`
2. Press `Cmd + Ctrl + Z` to open Developer Menu
3. Select `Toggle Element Inspector`
4. Try to select `Item 1`

Expected:
`Item 1` should be highlighted

Actual:
`Item 3` is hightlighted

### React Native Version

Any Version that has Fabric

### Affected Platforms

Runtime - Android, Runtime - iOS

### Areas

Fabric - The New Renderer

### Output of `npx react-native info`

```text
> ReproducerApp@0.0.1 npx
> react-native info

info Fetching system and libraries information...
System:
  OS: macOS 14.5
  CPU: (16) arm64 Apple M3 Max
  Memory: 7.91 GB / 128.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.12.2
    path: ~/.nvm/versions/node/v20.12.2/bin/node
  Yarn:
    version: 3.6.4
    path: ~/.nvm/versions/node/v20.12.2/bin/yarn
  npm:
    version: 10.6.0
    path: ~/.nvm/versions/node/v20.12.2/bin/npm
  Watchman:
    version: 2024-07-14T04:03:59Z
    path: /usr/local/bin/watchman
Managers:
  CocoaPods:
    version: 1.15.2
    path: /opt/homebrew/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.5
      - iOS 17.5
      - macOS 14.5
      - tvOS 17.5
      - visionOS 1.2
      - watchOS 10.5
  Android SDK:
    API Levels:
      - "26"
      - "27"
      - "28"
      - "29"
      - "30"
      - "31"
      - "33"
      - "34"
    Build Tools:
      - 33.0.1
      - 34.0.0
    System Images:
      - android-26 | Google APIs ARM 64 v8a
      - android-34 | Google APIs ARM 64 v8a
    Android NDK: Not Found
IDEs:
  Android Studio: 2023.1 AI-231.9392.1.2311.11330709
  Xcode:
    version: 15.4/15F31c
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 11.0.22
    path: /Users/soelynn/.sdkman/candidates/java/current/bin/javac
  Ruby:
    version: 2.6.10
    path: /usr/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.74.3
    wanted: 0.74.3
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: true
```

### Stacktrace or Logs

```text
N/A
```

### Reproducer

https://github.com/realsoelynn/rn-inspectortools-transform-related-reproducer

### Screenshots and Videos

https://github.com/user-attachments/assets/86f083ca-9a4d-40ef-a7f2-72da037ef4f2

### PR DESCRIPTION
**Pull request by zhongwuzw**
## Summary:

FIxes #45502 .  cc @realsoelynn 

## Changelog:

[GENERAL] [FIXED] - Fixes findNodeAtPoint when views were inverted

## Test Plan:

Demo in #45502 .

### CODE REVIEW

**Comment by realsoelynn** [comment]
Thank you so much for quickly looking into the fix for this

Since this is relying on scaleX and scaleY, it will have issue when rotate is involved. (I added a new test case in the reproducer app as shown in the screenshot)



I am thinking of consolidating this `findNodeAtPoint` with [computeRelativeLayoutMetrics](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactCommon/react/renderer/core/LayoutableShadowNode.cpp#L59C37-L175) as that is being used in measure function to know where in the screen it's being rendered as findNodeAtPoint has been a major area for bugs due to this two different approaches.

**Comment by zhongwuzw** [comment]
I'm not familiar with measurement, will find some time to try to figure it out :)

**Comment by zhongwuzw** [comment]
@realsoelynn Hi, sorry I have no solution to consolidate them right now. But I tried to fix your rotated repro demo :) .

**Comment by realsoelynn** [comment]
C++ Test for findNodeAtPoint

Just dumping this test case i wrote a while back. Please help me add to the PR

```c++
TEST(FindNodeAtPointTest, invertedList) {
  auto builder = simpleComponentBuilder();
  
  // clang-format off
  auto element =
    Element<ScrollViewShadowNode>()
      .props([] {
        auto sharedProps = std::make_shared<ScrollViewProps>();
        sharedProps->transform = Transform::VerticalInversion();
        return sharedProps;
      })
      .tag(1)
      .finalize([](ScrollViewShadowNode &shadowNode){
        auto layoutMetrics = EmptyLayoutMetrics;
        layoutMetrics.frame.size = {100, 200};
        shadowNode.setLayoutMetrics(layoutMetrics);
      })
      .children({
        Element<ViewShadowNode>()
        .tag(2)
        .finalize([](ViewShadowNode &shadowNode){
          auto layoutMetrics = EmptyLayoutMetrics;
          layoutMetrics.frame.origin = {0, 0};
          layoutMetrics.frame.size = {100, 100};
          shadowNode.setLayoutMetrics(layoutMetrics);
        }),
        Element<ViewShadowNode>()
        .tag(3)
        .finalize([](ViewShadowNode &shadowNode){
          auto layoutMetrics = EmptyLayoutMetrics;
          layoutMetrics.frame.origin = {0, 100};
          layoutMetrics.frame.size = {100, 100};
          shadowNode.setLayoutMetrics(layoutMetrics);
        })
    });
  // clang-format on

  auto parentShadowNode = builder.build(element);

  EXPECT_EQ(
      LayoutableShadowNode::findNodeAtPoint(parentShadowNode, {10, 10})
          ->getTag(),
      3);
  EXPECT_EQ(
      LayoutableShadowNode::findNodeAtPoint(parentShadowNode, {10, 105})
          ->getTag(),
      2);
}
```

**Comment by realsoelynn** [comment]
@zhongwuzw Please take a look at this other author PR https://github.com/facebook/react-native/pull/45571

Your current solution handled perfectly for horizontal and vertical inversion.

But, we want to to be able to handle different type of rotation transformation.

https://github.com/facebook/react-native/pull/45571 I don't agree with the Inversion check in that PR. But, the added new Point transformation function has the right idea to solve rotational transform

**Comment by coado** [comment]
@realsoelynn Is there any way to run cpp tests on emulator locally?

**Comment by coado** [comment]
> @zhongwuzw Please take a look at this other author PR #45571
> 
> Your current solution handled perfectly for horizontal and vertical inversion.
> 
> But, we want to to be able to handle different type of rotation transformation.
> 
> #45571 I don't agree with the Inversion check in that PR. But, the added new Point transformation function has the right idea to solve rotational transform
> 
> Cc: @coado

Do you mean this [inversion check](https://github.com/facebook/react-native/pull/45571/files#diff-5d2aefcdf4e9ea500a3053424efe5f4b02cad863c3af36dd9cb8d18a5a925403R272-R274)?

**Comment by zhongwuzw** [comment]
Maybe we can combine the two PRs: one for a quick path to handle horizontal and vertical inversion, and another for a transform fallback to matrix inversion calculation ？

**Comment by realsoelynn** [comment]
> > @zhongwuzw Please take a look at this other author PR #45571
> > Your current solution handled perfectly for horizontal and vertical inversion.
> > But, we want to to be able to handle different type of rotation transformation.
> > #45571 I don't agree with the Inversion check in that PR. But, the added new Point transformation function has the right idea to solve rotational transform
> > Cc: @coado
> 
> Do you mean this [inversion check](https://github.com/facebook/react-native/pull/45571/files#diff-5d2aefcdf4e9ea500a3053424efe5f4b02cad863c3af36dd9cb8d18a5a925403R272-R274)?

I mean `bool isInverted = currentTransform.inv();`. `inv()` function is not only returning bool result for if it's inverted, but also mutating the current transform.`inv()` function should not mutate.

**Comment by realsoelynn** [comment]
> Maybe we can combine the two PRs: one for a quick path to handle horizontal and vertical inversion, and another for a transform fallback to matrix inversion calculation ？

@zhongwuzw  yep. do you mind pulling in @coado commit to this PR?

File: packages/react-native/ReactCommon/react/renderer/graphics/Transform.cpp [hunk 1]
```diff
@@ -487,4 +498,74 @@ Size operator*(const Size& size, const Transform& transform) {
...
+  if (det == 0)
+    return false;
+
+  det = 1.0 / det;
+
+  for (i = 0; i < 16; i++)
+    this->matrix[i] = inv[i] * det;
+
+  return true;
+}
```
Comment (realsoelynn) [inline packages/react-native/ReactCommon/react/renderer/graphics/Transform.cpp:569]: i don't think we need this.

**Comment by coado** [comment]
I think right now changes are merged in a way that omits handling rotations. I don't see my part of the code in `findNodeAtPoint` function, that's why the `inv` is not being invoked. Also, does it make sense to handle horizontal and vertical inversion separately? The inversion matrix will handle all cases.

**Comment by zhongwuzw** [comment]
@coado Because of conflict I just merged some parts of the code from you, after I combine them, I'll push the whole code. Also I would remove the overflow part which we can concentrate on transform things.

**Comment by zhongwuzw** [comment]
cc @realsoelynn @coado

**Comment by realsoelynn** [comment]
Sorry about my late reply

It's failing UT on `FindNodeAtPointTest/viewIsScaled` test case

```sh
Failure: Expected equality of these values:
  LayoutableShadowNode::findNodeAtPoint(parentShadowNode, {119, 119})->getTag()
    Which is: 3
  2
```

I am figuring out how to run C++ Unit Test like above to run without BUCK. (I will keep you folks posted about it once i learn more about it.)

File: packages/react-native/ReactCommon/react/renderer/graphics/Transform.cpp [hunk 2]
```diff
@@ -487,4 +498,70 @@ Size operator*(const Size& size, const Transform& transform) {
...
+  inv[7] = m[0] * m[6] * m[11] - m[0] * m[7] * m[10] - m[4] * m[2] * m[11] +
+      m[4] * m[3] * m[10] + m[8] * m[2] * m[7] - m[8] * m[3] * m[6];
+
+  inv[11] = -m[0] * m[5] * m[11] + m[0] * m[7] * m[9] + m[4] * m[1] * m[11] -
+      m[4] * m[3] * m[9] - m[8] * m[1] * m[7] + m[8] * m[3] * m[5];
+
+  inv[15] = m[0] * m[5] * m[10] - m[0] * m[6] * m[9] - m[4] * m[1] * m[10] +
+      m[4] * m[2] * m[9] + m[8] * m[1] * m[6] - m[8] * m[2] * m[5];
+
+  det = m[0] * inv[0] + m[1] * inv[4] + m[2] * inv[8] + m[3] * inv[12];
```
Comment (realsoelynn) [inline packages/react-native/ReactCommon/react/renderer/graphics/Transform.cpp:554]: @coado Can i have a link to details explanation about this matrix calculation?

I would like to understand more about it.

Comment (coado) [reply packages/react-native/ReactCommon/react/renderer/graphics/Transform.cpp:554]: I've opened a draft, just to check if the idea is correct. So, for now I just copied the implementation from the [stackoverflow](https://stackoverflow.com/questions/1148309/inverting-a-4x4-matrix). There is a more detailed implementation of this in mesa library [here](https://gitlab.freedesktop.org/mesa/mesa/-/blob/main/src/util/u_math.c#L139-311).

**Comment by coado** [comment]
> > Sorry about my late reply
> > It's failing UT on `FindNodeAtPointTest/viewIsScaled` test case
> > ```shell
> > Failure: Expected equality of these values:
> >   LayoutableShadowNode::findNodeAtPoint(parentShadowNode, {119, 119})->getTag()
> >     Which is: 3
> >   2
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am figuring out how to run C++ Unit Test like above to run without BUCK. (I will keep you folks posted about it once i learn more about it.)
> 
> Seems this test is wrong, the tag should be 3, not 2 ?

Isn't the view with tag 3 scaled down? It seems like the touch point is within the boundaries of unscaled view with tag 2, so the test is correct, but I might be missing something.

**Comment by zhongwuzw** [comment]
I used copilot to translate C++ test code to React code,

**Comment by zhongwuzw** [comment]
@realsoelynn I just translate C++ ut to React code like below: I hit the breakpoint and it returns `child1`.
```
import React from 'react';
import { View, StyleSheet } from 'react-native';

const FindNodeAtPointTest = () => {
  return (
    <View style={styles.parent}>
      <View style={styles.child1}>
        <View style={styles.child2} />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  parent: {
    width: 1000,
    height: 1000,
  },
  child1: {
    position: 'absolute',
    top: 100,
    left: 100,
    width: 100,
    height: 100,
    backgroundColor: 'blue'
  },
  child2: {
    position: 'absolute',
    top: 10,
    left: 10,
    width: 10,
    height: 10,
    backgroundColor: 'red',
    transform: [{ scale: 0.5 }],
  },
});

export default FindNodeAtPointTest;
```

image

**Comment by realsoelynn** [comment]
@zhongwuzw sorry about my delay. Let's go with your original solution for now to do a simple inversion (without @coado changes). As `0.76` cutoff is near, we want to at least get this major usecase bugfix went in first.

@coado For your changes, let me work with you more later after this PR is merged as i need to prioritize and research into the algorithms more. 

Thank you both for your patient.

File: packages/react-native/ReactCommon/react/renderer/core/LayoutableShadowNode.cpp [hunk 3]
```diff
@@ -261,10 +261,31 @@ ShadowNode::Shared LayoutableShadowNode::findNodeAtPoint(
...
+    if (Transform::isVerticalInversion(transform)) {
+      relativeY = -relativeY;
+    }
+    if (Transform::isHorizontalInversion(transform)) {
+      relativeX = -relativeX;
+    }
+
+    point.x = centerX + relativeX;
+    point.y = centerY + relativeY;
+  }
```
[the remark this answers is not in the document]
Comment (zhongwuzw) [reply packages/react-native/ReactCommon/react/renderer/core/LayoutableShadowNode.cpp:294]: @realsoelynn Done.

**Comment by realsoelynn** [comment]
Hmm strange CircleCI failure. @zhongwuzw please help us rebase it.

**Comment by zhongwuzw** [comment]
@realsoelynn Done. :)

**Pull request closed without merging**